File: examples/Python/flserver3.py
"""Freelance server - Model 3

Uses an ROUTER/ROUTER socket but just one thread

Author: Min RK <benjaminrk@gmail.com>
"""

import sys

import zmq

from zhelpers import dump

def main():
    verbose = '-v' in sys.argv
    
    ctx = zmq.Context()
    # Prepare server socket with predictable identity
    bind_endpoint = "tcp://*:5555"
    connect_endpoint = "tcp://localhost:5555"
    server = ctx.socket(zmq.ROUTER)
    server.identity = connect_endpoint
    server.bind(bind_endpoint)
    print "I: service is ready at", bind_endpoint
    
    while True:
        try:
            request = server.recv_multipart()
        except:
            break # Interrupted
        # Frame 0: identity of client
        # Frame 1: PING, or client control frame
        # Frame 2: request body
        address, control = request[:2]
        reply = [address, control]
        if control == "PING":
            reply[1] = "PONG"
        else:
            reply.append("OK")
        if verbose:
            dump(reply)
        server.send_multipart(reply)
    print "W: interrupted"

if __name__ == '__main__':
    main()

File: examples/Lua/flserver3.lua
--
--  Freelance server - Model 3
--  Uses an ROUTER/ROUTER socket but just one thread
--
--  Author: Robert G. Jakabosky <bobby@sharedrealm.com>
--
require"zmq"
require"zmsg"

local verbose = (arg[1] == "-v")

local context = zmq.init(1)
s_catch_signals ()

--  Prepare server socket with predictable identity
local bind_endpoint = "tcp://*:5555"
local connect_endpoint = "tcp://localhost:5555"
local server = context:socket(zmq.ROUTER)
server:setopt(zmq.IDENTITY, connect_endpoint)
server:bind(bind_endpoint)
printf ("I: service is ready at %s\n", bind_endpoint)

while (not s_interrupted) do
    local request = zmsg.recv (server)
    local reply = nil
    if (not request) then
        break          --  Interrupted
    end
    if (verbose) then
        request:dump()
    end
    --  Frame 0: identity of client
    --  Frame 1: PING, or client control frame
    --  Frame 2: request body
    local address = request:pop()
    if (request:parts() == 1 and request:body() == "PING") then
        reply = zmsg.new ("PONG")
    elseif (request:parts() > 1) then
        reply = request
        request = nil
        reply:body_set("OK")
    end
    reply:push(address)
    if (verbose and reply) then
        reply:dump()
    end
    reply:send(server)
end
if (s_interrupted) then
    printf ("W: interrupted\n")
end
server:close()
context:term()


